castai/resource_autoscaler.go

```diff
@@ -49,12 +89,381 @@ func resourceAutoscaler() *schema.Resource {
+												},
+											},
+										},
+									},
+									FieldNodeConstraints: {
+										Type:        schema.TypeList,
+										Optional:    true,
+										MaxItems:    1,
+										Description: "defines the node constraints that will be applied when autoscaling with Unschedulable Pods policy.",
+										Elem: &schema.Resource{
+											Schema: map[string]*schema.Schema{
+												FieldMinCPUCores: {
+													Type:        schema.TypeInt,
+													Optional:    true,
+													Default:     0,
+													Description: "defines min CPU cores for the node to pick.",
+												},
+												FieldMaxCPUCores: {
+													Type:        schema.TypeInt,
+													Optional:    true,
+													Default:     32,
+													Description: "defines max CPU cores for the node to pick.",
+												},
+												FieldMinRAMMiB: {
+													Type:        schema.TypeInt,
+													Optional:    true,
+													Default:     2048,
+													Description: "defines min RAM in MiB for the node to pick.",
+												},
+												FieldMaxRAMMiB: {
+													Type:        schema.TypeInt,
+													Optional:    true,
+													Default:     262144,
+													Description: "defines max RAM in MiB for the node to pick.",
+												},
+												FieldEnabled: {
+													Type:        schema.TypeBool,
+													Optional:    true,
+													Default:     false,
+													Description: "enable/disable node constraints policy.",
+												},
+											},
+										},
+									},
+									FieldCustomInstancesEnabled: {
+										Type:        schema.TypeBool,
+										Optional:    true,
+										Default:     false,
+										Description: "enable/disable custom instances policy.",
+									},
+								},
+							},
+						},
+						FieldClusterLimits: {
+							Type:        schema.TypeList,
+							Optional:    true,
+							MaxItems:    1,
+							Description: "defines minimum and maximum amount of CPU the cluster can have.",
+							Elem: &schema.Resource{
+								Schema: map[string]*schema.Schema{
+									FieldEnabled: {
+										Type:        schema.TypeBool,
+										Optional:    true,
+										Default:     true,
+										Description: "enable/disable cluster size limits policy.",
+									},
+									FieldCPU: {
+										Type:        schema.TypeList,
+										Optional:    true,
+										MaxItems:    1,
+										Description: "defines the minimum and maximum amount of CPUs for cluster's worker nodes.",
+										Elem: &schema.Resource{
+											Schema: map[string]*schema.Schema{
+												FieldMinCores: {
+													Type:             schema.TypeInt,
+													Optional:         true,
+													Default:          1,
+													Description:      "defines the minimum allowed amount of CPUs in the whole cluster.",
+													ValidateDiagFunc: validation.ToDiagFunc(validation.IntBetween(1, 20)),
+												},
```

Comment:
why only 1 to 20 when max can be 2 to 1000?

Reply:
These values are based on our api contract definitons

```proto
// Defines the minimum and maximum amount of vCPUs for cluster's worker nodes.
message ClusterLimitsCpu {
  // Defines the minimum allowed amount of CPUs in the whole cluster.
  int32 min_cores = 1 [(grpc.gateway.protoc_gen_openapiv2.options.openapiv2_field) = { default: "1", minimum: 1, maximum: 20, example: "1" }];
  // Defines the maximum allowed amount of vCPUs in the whole cluster.
  int32 max_cores = 2 [(grpc.gateway.protoc_gen_openapiv2.options.openapiv2_field) = { default: "20", minimum: 2, maximum: 1000, example: "20" }];
}
```

https://gitlab.com/castai/kubecast/-/blob/master/api-contracts/grpc/proto/policies/v1/policy.proto?ref_type=heads#L110